src/libcore/str/mod.rs

```diff
@@ -34,99 +36,33 @@ use result::Result::{self, Ok, Err};
 use slice::{self, SliceExt};
 use usize;
 
-pub use self::pattern::Pattern;
-pub use self::pattern::{Searcher, ReverseSearcher, DoubleEndedSearcher, SearchStep};
+pub mod pattern;
 
-mod pattern;
+macro_rules! iter_methods {
+    (: $te:ty) => {
```

Comment:
In general, I tend to use macros like these a lot if I recognize a common pattern across many different places - in this case the pattern of delegating a API to a field of the struct.

In this case, it could probably be inlined without too much source bloat, but the the manual impl approach has the problem that it tends to miss things like default methods. 

I agree though that as its right now it doesn't quite pull its weight, do you think a more explicit syntax like `iter_methods!(delegate Iterator to self.0)` would solve the understandability issue?

(It would also help if we could write macros that can work with lifetime and type bound lists, as that would allow to write a single common macro that could be used across different modules for this task...)